server/fileupdater.mjs

```diff
@@ -7,56 +7,106 @@ export default function FileUpdater(state) {
   if(v > VERSION)
     throw Error(`File version ${v} is newer than the supported version ${VERSION}.`);
 
+  const globalProperties = computeGlobalProperties(state, v);
   for(const id in state)
-    updateProperties(state[id], v);
+    updateProperties(state[id], v, globalProperties);
 
   state._meta.version = VERSION;
   return state;
 }
 
-function updateProperties(properties, v) {
+function computeGlobalProperties(state, v) {
+  let globalProperties = {};
+  if (globalProperties.v12DropShadowAllowed = v < 10) {
+    for (const id in state) {
+      const properties = state[id];
+      if (properties.type == 'card' || properties.type == 'deck' || properties.type == 'pile') {
+        globalProperties.v12DropShadowAllowed = globalProperties.v12DropShadowAllowed &&
+            !hasPropertyCondition(properties, (properties) => {
+              return properties.parentChangeRoutine || properties.changeRoutine;
+            });
+      }
+      globalProperties.v12DropShadowAllowed = globalProperties.v12DropShadowAllowed &&
+          !hasPropertyCondition(properties, (properties) => {
+            return properties.parentGlobalUpdateRoutine || properties.globalUpdateRoutine;
+          });
+      if (!globalProperties.v12DropShadowAllowed)
+        break;
+    }
+  }
+  return globalProperties;
+}
+
+function hasPropertyCondition(properties, condition) {
+  if (typeof properties != 'object')
+    return false;
+  if (condition(properties))
+    return true;
+  if (properties.type) {
+    if (hasPropertyCondition(properties.onPileCreation, condition))
+      return true;
+    if (typeof properties.faces == 'object') {
+      for (const face of properties.faces) {
+        if (hasPropertyCondition(properties.faces[face], condition))
+          return true;
+      }
+    }
+    if (properties.type == 'deck') {
+      if (hasPropertyCondition(properties.cardDefaults, condition))
+        return true;
+      if (typeof properties.cardTypes == 'object')
+        for(const cardType in properties.cardTypes)
+          if (hasPropertyCondition(properties.cardTypes[cardType], condition))
+            return true;
+    }
```

Comment:
This seems to be missing `faceTemplates.x.properties`.

Reply:
As far as I can tell, faceTemplates properties don't ever set properties on the widget so things like `parentChangeRoutine` would be irrelevant on those wouldn't it?

I also noticed updateProperties doesn't update those either - though there have been specific version updates which do things to face templates.

Reply:
> As far as I can tell, faceTemplates properties don't ever set properties on the widget so things like `parentChangeRoutine` would be irrelevant on those wouldn't it?

They "set" properties on the cards just like `cardDefaults` or `cardTypes`. Note that the properties are inside a property called `properties` in the face definition.

> I also noticed updateProperties doesn't update those either - though there have been specific version updates which do things to face templates.

I thought, I saw them yesterday.

Reply:
Thanks, it wasn't very obvious to me where this happened but I see now the properties are applied in the card's `applyDeltaToDOM` function. I've updated the fileupdater logic to check this.